packages/eslint-plugin-jsx/src/rules/no-leaked-conditional-rendering.ts

```diff
@@ -206,7 +206,14 @@ export default createRule<[], MessageID>({
     function isValidConditionalExpression(
       node: TSESTree.ConditionalExpression,
     ): boolean {
-      const { alternate, consequent } = node;
+      const { alternate, consequent, test } = node;
```

Comment:
In this scenario we don't care about the value of the test node at all, whatever it is will never be inserted into JSX and rendered by react (only consequent and alternate will).

Reply:
What I'm thinking about is that I might just want to render this number when it's non-null or non-0.

```
x ? x : fallback
```

Reply:
> What I'm thinking about is that I might just want to render this number when it's non-null or non-0.
> 
> ```
> x ? x : fallback
> ```

That will not work, in the updated test case you have annotate the type to someCondition?: number | undefined, which include falsy number 0.

Only when it's type is non-zero number literal types like 2, 3, 5... make sense
e.g. someCondition?: 5 | undefined, and I can achieve this with a single code change, but it's not to add the test checking.

Reply:
> > What I'm thinking about is that I might just want to render this number when it's non-null or non-0.
> > ```
> > x ? x : fallback
> > ```
> 
> That will not work, in the updated test case you have annotated the type to someCondition?: number | undefined, which include falsy number 0.
> 
> Only when it's type is non-zero number literal types like 2, 3, 5... make sense e.g. someCondition?: 5 | undefined, and the code added for this case has not change the type of the literal before return true (not checked non-literal may coverted using valueOf before the test comparisons which we can't tell if it's safe).

Some information was incorrect and has been updated.

Reply:
I apologize for the inaccuracy of the initial response.
I took a look at it, We can indeed check if the `consequent` is the same as `test`!

Add
```ts
const allowGuardedTypes = [
  "boolean",
  "string",
  "number",
  "nullish",

  "truthy boolean",
  "truthy number",
  "truthy string",
] as const;
```

And Then change the `isValidConditionalExpression` function into this:

```
function isValidConditionalExpression(
  node: TSESTree.ConditionalExpression,
): boolean {
  const { alternate, consequent, test } = node;

  const isConsequentGuarded = isNodeEqual(consequent, test);
  const isAlternateGuarded = isNodeEqual(alternate, test);
  const testType = getConstrainedTypeAtLocation(services, test);
  const types = inspectVariantTypes([testType]);

  if (isConsequentGuarded && types.every(type => allowGuardedTypes.includes(type as never))) {
    return true;
  }

  if (isAlternateGuarded && types.every(type => allowGuardedTypes.includes(type as never))) {
    return true;
  }

  return isValidExpression(alternate) && isValidExpression(consequent);
}
```

Reply:
It looks like `inspectVariantTypes([testType])` doesn't get the correct type for test case `someCondition?: number`. Because it is a union type, we need to pass `testType.types`.
```
types === ['object']
```